rust/src/instruction/declaration.rs

```diff
@@ -0,0 +1,145 @@
+use pyo3::prelude::*;
+use pyo3::types::PyDict;
+
+use crate::conversion::{optional_tuples_from_offsets, string_from_scalar_type};
+use crate::instruction::Instruction;
+use crate::{conversion, extract_instruction_as, extract_instruction_as_mut, impl_from_quil_rs};
+
+#[pyclass(extends=Instruction)]
+#[derive(Debug, Clone)]
+pub struct Declare {}
+impl_from_quil_rs!(Declare, quil_rs::instruction::Declaration, Declaration);
+
+#[pymethods]
+impl Declare {
+    #[new]
+    #[pyo3(signature=(name, memory_type, memory_size=1, shared_region=None, offsets=Vec::new()))]
+    pub fn new(
+        name: String,
+        #[pyo3(from_py_with = "conversion::py_to_scalar_type")]
+        memory_type: quil_rs::instruction::ScalarType,
+        memory_size: u64,
+        shared_region: Option<String>,
+        #[pyo3(from_py_with = "conversion::py_to_offsets")] offsets: Vec<
+            quil_rs::instruction::Offset,
+        >,
+    ) -> PyResult<(Self, Instruction)> {
```

Comment:
In order to return a class that extends another, you need to return a tuple containing the extension class and the base class (or a `PyClassInitializer` for more complex cases. See [pyo3 inheritance](https://pyo3.rs/v0.21.0-beta.0/class.html?highlight=subclass#inheritance).

As for `Declare` and all the other concrete instruction classes, they are empty because the `Instruction` class is what contains the actual `quil_rs::Instruction`, so there isn't anything relevant to store in the child classes.